### PR TITLE
Features/annotation routing

### PR DESCRIPTION
**Pull request by siarhei-baradzin**
https://oat-sa.atlassian.net/browse/TAO-8515

It's a draft of proposed routing improvement which solves issues listed in the task linked above. Depends on https://github.com/oat-sa/generis/pull/637

On the current stage, I need **only high-level architectural review**. Probably contains some naming/code style/lack of comments issues.

The main idea is to add support of **`@route`** annotation in controllers with full backward compatibility with existing controllers and methods. 

To enable annotations support we need to use a new router in manifest.php (tested locally with new caed extension):
```php
...
'routes' => array(
        '/taoCaedWorkflow' => [
            'class' => \oat\tao\model\routing\ControllerAnnotationsRoute::class,
            'namespace' => 'oat\\taoCaedWorkflow\\controller'
        ]
    ),  
...
```
This router compatible with default NamespaceRouter but also handles annotated methods. Example of a controller with public methods annotated with `@route` :
```php
namespace oat\taoCaedWorkflow\controller;

use oat\tao\model\routing\AnnotationReader\route;

class Assignee extends \tao_actions_CommonModule
{
    public function oldStyle() {
        $this->returnJson(['ddd' => 'yep']);
    }

    public function index() {
        $this->returnJson(['index' => 'yep']);
    }

    /**
     * @route(method="GET", relativePath="{assigneeId}/tasks")
     * @param string $assigneeId
     */
    public function tasks($assigneeId) {
        $this->returnJson(['passed_a_id' => $assigneeId]);
    }

    /**
     * @route(method="POST", relativePath="ccc")
     */
    public function bbb() {
        $this->returnJson(['bbb' => 'yep']);
    }

    public function ccc() {
        $this->returnJson(['ccc' => 'yep']);
    }
}
```
1. Methods without annotation are still available for any HTTP method as it was before: `/taoCaedWorflow/Assignee/oldStyle`. 
2. `index` is still available by `/taoCaedWorkflow/Assignee` or `/taoCaedWorkflow/Assignee/index`
3. method `tasks` is available by path `GET /taoCaedWorkflow/Assignee/{assigneeId}/tasks` where `{assigneeId}` gets extracted and added to request params (and therefore passed as argument)
4. method `bbb` is available by path `POST /taoCaedWorkflow/Assignee/ccc`
5. Method `ccc` is not available anymore because `bbb` overrides it's default path by annotation
6. Request existing route with wrong HTTP method produces correct 405 (Method Not Allowed) answer (no need to check manually in each method)
7. Route annotation params: 
    * `method` is optional, by default any HTTP method allowed.
    * `relativePath` is optional, by default method name is used

Implementation details:
1. `route` annotation class added to existing `requiredRights` and `security` annotation
2. `AnnotationReaderService` modifications:
    * Support of parsing all public methods annotations for particular class added
    * Optional param to parse only specified annotation
    * Option to disable annotations cache (for development)
3. new `interface PathVariablesProvidingRoute extends Route` describes custom route which can parse variables from a path. `ControllerAnnotationsRoute` implements it, old `NamespaceRoute` doesn't 
4. `ActionEnforser` and `Resolver` modified to add such variables to request params
5. `ExceptionInterpretor` was modified to handle `common_exception_MethodNotAllowed` in a proper way (405 response with 'Allow' header, html teplate added)
6. Draft implementation of `DeleteRoutesCache` CLI command added (for developers)
7. Only already existing 3-rd part libs are used: Slim, FastRoute. I tried to use existing code and make changes as small as I can, that's why only routing on controllers level was implemented.

Possible improvements: use FastRoutes's level cache (with prepared regexps) in separate files instead of CacheService (performance, easier to clear)

### CODE REVIEW

**Comment by julien-sebire** [comment]
Hi Siarhei,
Good job!
Three tests are failing (details: http://qa.kitchen.it.taocloud.org/blue/organizations/jenkins/Github%20Pull%20Requests%20for%20oat-sa%2Ftao-core/detail/features%2FannotationRouting/1/pipeline/51).

**Review by hutnikau — CHANGES_REQUESTED** [review]

**Comment by siarhei-baradzin** [comment]
> Hi Siarhei,
> Good job!
> Three tests are failing (details: http://qa.kitchen.it.taocloud.org/blue/organizations/jenkins/Github%20Pull%20Requests%20for%20oat-sa%2Ftao-core/detail/features%2FannotationRouting/1/pipeline/51).

Thx, fixed

File: models/classes/routing/ControllerAnnotationsRoute.php [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2014 (original work) Open Assessment Technologies SA;
+ *
+ *
+ */
+namespace oat\tao\model\routing;
+
+use FastRoute\Dispatcher;
```
[the remark this answers is not in the document]
Comment (siarhei-baradzin) [reply models/classes/routing/ControllerAnnotationsRoute.php]: it's an essential dependency of 'slim' which is already required in composer (they even expose FastRoute consts in documentation), I use only these consts from it

File: models/classes/routing/Resolver.php [hunk 2]
```diff
@@ -236,6 +258,11 @@ private function getRoute(\common_ext_Extension $extension, $routeId, $routeData
             throw new \common_exception_InconsistentData('Invalid route '.$routeId);
         }
         $className = $routeData['class'];
-        return new $className($extension, trim($routeId, '/'), $routeData);
+        /** @var Route $route */
+        $route = new $className($extension, trim($routeId, '/'), $routeData);
+        if ($route instanceof ServiceLocatorAwareInterface) {
+            $route->setServiceLocator($this->getServiceLocator());
```
[the remark this answers is not in the document]
Comment (siarhei-baradzin) [reply models/classes/routing/Resolver.php]: We don't have `$this->getServiceManager()` here

**Pull request closed without merging**